controller: Refer port by profile name

cathay4t opened the pull request:
This allows controller port referring to the profile name:

```yml
---
interfaces:
- name: port1
  type: ethernet
  identifier: mac-address
  mac-address: 00:23:45:67:89:1a
- name: port2
  type: ethernet
  identifier: mac-address
  mac-address: 00:23:45:67:89:1b
- name: br0
  type: linux-bridge
  state: up
  bridge:
    ports:
      - name: port1
      - name: port2
```

This patch enable this support to bond, vrf, linux-bridge and
ovs-bridge.

Crated unit tests and integration tests for these use cases:
 * Refer port by mac address for bond/linux bridge/ovs bridge/vrf
 * Prefer kernel name over profile name when overlap.
 * Raise error if controller refer to the profile name shared by 
   multiple interfaces.
 * Allow multiple interfaces sharing the same profile name as long as
   they are not referred by controller.


Resolves: https://issues.redhat.com/browse/RHEL-1745

cathay4t commented:
Improvements comparing to previous solutions:

1. No changes to schema.
2. Better align with user's intuition
3. Way less code changes comparing to #2710

cathay4t commented:
This PR does not have `gen_conf` support. Once this schema is accepted, I will create new PR for `gen_conf` support on this.

ihuguet commented on rust/src/lib/ifaces/bond.rs:
```diff
@@ -374,9 +374,21 @@ impl BondInterface {
                 .and_then(|bond_conf| bond_conf.port.as_mut())
             {
-                ports[index] = new_name;
+                ports[index] = new_name.clone();
             }
         }
+        if let Some(port_conf) = self
+            .bond
+            .as_mut()
+            .and_then(|bond_conf| bond_conf.ports_config.as_mut())
+            .and_then(|port_confs| {
+                port_confs
+                    .iter_mut()
+                    .find(|port_conf| port_conf.name == origin_name)
+            })
+        {
+            port_conf.name = new_name;
+        }
```
It is not allowed to have `port` and `port_conf` at the same time, isn't it? Then, you can do an `if port {...} else if ports_config {...}`. It would avoid `new_name.clone()`, saving an allocation.

cathay4t replied:
It is allowed to have both defined at the same time because that is how nmstate query outputs.

Nmstate should allows such when they are not conflict with each other.

ihuguet commented on rust/src/lib/ifaces/inter_ifaces_controller.rs:
```diff
@@ -157,6 +157,61 @@ impl MergedInterfaces {
+            }
+        }
+        // TODO: Validate profile name conflict with other interfaces' kernel
+        // name
```
Maybe create first a HashSet with all the interface names. Then, when filling `profile_2_kernel`, check that `profile_name` is not in that list, and ignore it if it is..

cathay4t replied:
I have added `MergedInterfaces::validate_profile_name_conflict()` .

Both unit test case and integration test case are included for this invalid desired state.

cathay4t replied:
This validation has removed in order to behave exactly as before.

ihuguet commented on rust/src/lib/ifaces/vrf.rs:
```diff
@@ -126,6 +126,29 @@ impl VrfInterface {
+            .as_ref()
+            .and_then(|vrf_conf| vrf_conf.port.as_ref())
+            .and_then(|ports| {
+                ports.iter().position(|port_name| port_name == origin_name)
```
Use `find`  instead of `position` here and you will get the port directly, so you won't need the second `if let Some(...`

cathay4t replied:
~~The `find()` is read only.~~

Let me try it out.

cathay4t replied:
Changed to
```rust
        if let Some(port_name) = self
            .vrf
            .as_mut()
            .and_then(|vrf_conf| vrf_conf.port.as_mut())
            .and_then(|ports| {
                ports
                    .iter_mut()
                    .find(|port_name| port_name.as_str() == origin_name)
            })
        {
            *port_name = new_name;
        }
```

ihuguet reviewed (changes requested):
It would be good to add some tests using the profile name, and also some where the profile name conflicts with an interface name.

Overall, the change is good, I like it more than the previous approach. This one is much simpler, the syntax fits well with the existing style and from the user perspective is easy to use.

cathay4t commented:
New patches uploaded.

Added two more checks on invalid use cases:
 * 2+ interfaces holding the same name and type.
 * Profile name is conflicting with other kernel name.

ihuguet commented:
I wonder if using the field `name` can be confusing. Interfaces have the `profile_name` field, woudn't make more sense to use it? More intuitive and less ambiguous.

```
interfaces:
- profile-name: port1
  type: ethernet
  identifier: mac-address
  mac-address: 00:23:45:67:89:1a
- profile-name: port2
  type: ethernet
  identifier: mac-address
  mac-address: 00:23:45:67:89:1b
- name: br0
  type: linux-bridge
  state: up
  bridge:
    ports:
      - name: port1
      - name: port2
```

cathay4t commented:
Previous code already treat `interface.name` as profile name for mac-based identifier NIC.

Meanwhile, to support what you said require big changes to our deserializer of `Interfaces` which assume all interface has `interface.name`.

ihuguet commented on rust/src/lib/ifaces/inter_ifaces_controller.rs:
```diff
@@ -157,6 +157,59 @@ impl MergedInterfaces {
+                continue;
+            }
+            let ports: Vec<String> = if let Some(ports) = des_iface.ports() {
+                ports.iter().map(|p| p.to_string()).collect()
```
You don't need the `to_string` here. Let `ports` to be a `Vec<&str>`.

cathay4t replied:
```rust
                    des_iface
                        .change_port_name(&port_name, kernel_name.to_string());
```

The `des_iface` will be changed later, so you cannot hold reference to it.

ihuguet commented on rust/src/lib/ifaces/inter_ifaces_controller.rs:
```diff
@@ -157,6 +157,59 @@ impl MergedInterfaces {
+    // If port name reference should be check kernel name first, then fallback
+    // to profile name.
+    pub(crate) fn resolve_port_name_ref(&mut self) -> Result<(), NmstateError> {
+        let mut profile_2_kernel: HashMap<String, String> = HashMap::new();
+        for iface in self.kernel_ifaces.values() {
+            let base_iface = iface.merged.base_iface();
+            if let Some(profile_name) = base_iface.profile_name.as_deref() {
+                if profile_name != base_iface.name.as_str() {
+                    profile_2_kernel.insert(
+                        profile_name.to_string(),
+                        base_iface.name.to_string(),
+                    );
+                }
+            }
+        }
```
Probably you can use a `HashMap<&str, &str>` here?

cathay4t replied:
We will change `des_iface`, `for_apply` and `for_verify` of `self`, so we have to own the data instead of referencing.

ihuguet commented on rust/src/lib/ifaces/inter_ifaces.rs:
```diff
@@ -939,6 +950,34 @@ impl MergedInterfaces {
         Ok(())
     }
+
+    // Profile name should not conflict with other interfaces' kernel name.
+    fn validate_profile_name_conflict(&self) -> Result<(), NmstateError> {
```
I understand that it's better not to guess, and it's totally weird that the user has a profile name with the same name of another device. But I fear that this can cause regressions, raising errors on configurations that were working fine, previously.
I'm not really sure what the best thing to do is, but I think that it would be better to match by interface name first, and fallback to profile name if it doesn't exist. If the user expected that the changes are applied to the other interface, it's their fault for having such a weird configuration, and there would not be any bug on our side. The behaviour is also 100% predictable.
@ffmancera @liangwen12year thoughts?

cathay4t replied:
Good point on the regression.

I think it is reasonable to check interface name first because that is how old nmstate is doing.

Let me fix it up.

cathay4t replied:
I have fixed the code to prefer kernel name over profile name when overlap.

And also allows multiple interfaces holding the same profile name as long as they are not refered by controller.

Unit and integration test cases included.

cathay4t committed:
controller: Refer port by profile name

This allows controller port referring to the profile name:

```yml
---
interfaces:
- name: port1
  type: ethernet
  identifier: mac-address
  mac-address: 00:23:45:67:89:1a
- name: port2
  type: ethernet
  identifier: mac-address
  mac-address: 00:23:45:67:89:1b
- name: br0
  type: linux-bridge
  state: up
  bridge:
    ports:
      - name: port1
      - name: port2
```

This patch enable this support to bond, vrf, linux-bridge and
ovs-bridge.

Crated unit tests and integration tests for these use cases:
 * Refer port by mac address for bond/linux bridge/ovs bridge/vrf
 * Prefer kernel name over profile name when overlap.
 * Raise error if controller refer to the profile name shared by
   multiple interfaces.
 * Allow multiple interfaces sharing the same profile name as long as
   they are not referred by controller.

Resolves: https://issues.redhat.com/browse/RHEL-1745

Signed-off-by: Gris Ge <fge@redhat.com>

ihuguet reviewed (approved):
Not sure if it's possible to get repeated profile names, I think not. But anyway, better to deal with it, just in case.
Approving.
